python/mobitex_deframer.py

```diff
@@ -0,0 +1,366 @@
+    crc_bytes: bytes,
+    max_bit_flips: int,
+    callsign_ref: bytes = None,
+) -> tuple[bytes, int]:
+    """
+    Attempt to error-correct a callsign by flipping bits until CRC matches.
+
+    Args:
+        callsign_bytes: The callsign bytes to decode
+        crc_bytes: The CRC bytes to verify against
+        max_bit_flips: Maximum number of bits to flip during error correction
+        callsign_ref: Unused argument
+
+    Returns:
+        tuple containing:
+            - corrected callsign bytes
+            - number of bit errors corrected
+
+    Raises:
+        ValueError: If no valid callsign could be found
+                    within max_bit_flips bit flips
+    """
+    if max_bit_flips <= 0:
+        # Do not attempt error correction
+        computed_crc = crc16_ccitt_zero(bytes(callsign_bytes))
+        computed_crc_bytes = computed_crc.to_bytes(length=2)
+
+        if computed_crc_bytes == crc_bytes:
+            return callsign_bytes, 0
+        else:
+            if max_bit_flips == 0:
+                raise ValueError('Callsign CRC check failed.')
+            else:
+                # If max_bit_flips is negative, just accept invalid CRC
+                return callsign_bytes, -1
+
+    # Bit position 0... (N*8); -1 indicates "no flip"
+    all_bytes = callsign_bytes + crc_bytes
+    flip_positions = range(-1, len(all_bytes) * 8)
+
+    for flips in itertools.combinations(flip_positions, max_bit_flips):
+        # Create a modified copy with some bits flipped
+        mod_bytes = bytearray(all_bytes)
+        for i in flips:
+            if i == -1:
+                # No flip
+                continue
+            i_byte = i // 8
+            i_bit = i % 8
+
+            # Flip one bit
+            mod_bytes[i_byte] = mod_bytes[i_byte] ^ (1 << i_bit)
+
+        # Split back into callsign and CRC
+        mod_callsign = mod_bytes[: len(callsign_bytes)]
+        mod_crc = mod_bytes[len(callsign_bytes):]
+
+        # Check if modified callsign matches its CRC
+        computed_crc_bytes = crc16_ccitt_zero(mod_callsign).to_bytes(length=2)
+
+        if computed_crc_bytes == bytes(mod_crc):
+            bit_errors = sum([1 for x in flips if x != -1])
+            return bytes(mod_callsign), bytes(mod_crc), bit_errors
+
+    raise ValueError(
+        f"No valid callsign found after flipping up to {max_bit_flips} bits"
+    )
+
+
+def hamming_distance(a: bytes, b: bytes) -> int:
+    # Convert bytes to bit arrays
+    a_bits = np.unpackbits(np.frombuffer(a, dtype=np.uint8))
+    b_bits = np.unpackbits(np.frombuffer(b, dtype=np.uint8))
+
+    # Count differences between bit arrays
+    bit_errors = np.count_nonzero(a_bits != b_bits)
+
+    return bit_errors
+
+
+def decode_known_callsign(
+    callsign_bytes: bytes,
+    crc_bytes: bytes,
+    max_bit_flips: int,
+    callsign_ref: bytes,
+) -> tuple[bytes, int]:
+
+    # Decode known callsign
+    callsign_crc_ref = crc16_ccitt_zero(callsign_ref).to_bytes(length=2)
```

Comment:
According to [the docs](https://docs.python.org/3/library/stdtypes.html#int.to_bytes), the default value for the `byteorder` argument of `to_bytes()` was added in Python 3.11, so it looks like this will need an explicit `byteorder='big'` to work with Python < 3.11.